From 9934316c2e7627823dc62a9129d23b5aea0d15ff Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Tue, 16 Sep 2014 17:10:24 -0400
Subject: [usb] core: add sanity checks when using bInterfaceClass with new_id

Message-id: <1410887463-20674-553-git-send-email-dzickus@redhat.com>
Patchwork-id: 95657
O-Subject: [RHEL7 PATCH 552/591] usb: core: add sanity checks when using bInterfaceClass with new_id
Bugzilla: 1110939
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1110939

commit c63fe8f6ca3669f1d120ff70523e2911b9966574
Author: Wolfram Sang <wsa@the-dreams.de>
Date:   Fri Jan 10 19:36:41 2014 +0100

    usb: core: add sanity checks when using bInterfaceClass with new_id

    Check if that field is actually used and if so, bail out if it exeeds a
    u8. Make it also future-proof by not requiring "exactly three"
    parameters in new_id, but simply "more than two".

    Signed-off-by: Wolfram Sang <wsa@the-dreams.de>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/usb/core/driver.c b/drivers/usb/core/driver.c
index aa01a3d..ccc307c 100644
--- a/drivers/usb/core/driver.c
+++ b/drivers/usb/core/driver.c
@@ -62,7 +62,10 @@ ssize_t usb_store_new_id(struct usb_dynids *dynids,
  dynid->id.idVendor = idVendor;
  dynid->id.idProduct = idProduct;
  dynid->id.match_flags = USB_DEVICE_ID_MATCH_DEVICE;
- if (fields == 3) {
+ if (fields > 2 && bInterfaceClass) {
+  if (bInterfaceClass > 255)
+   return -EINVAL;
+
   dynid->id.bInterfaceClass = (u8)bInterfaceClass;
   dynid->id.match_flags |= USB_DEVICE_ID_MATCH_INT_CLASS;
  }
-- 
1.7.1